File: gradle/subprojects/numbers-utils/src/main/kotlin/com/app/numbers/utils/operations/FactorUtils.kt
package com.app.numbers.utils.operations

import java.math.BigInteger
import kotlin.random.Random

object FactorUtils {
    fun factor(n: BigInteger): List<BigInteger> {
        val factors = mutableListOf<BigInteger>()
        val max = n.toLong()

        for (i in 1..max) {
            val divisor = BigInteger.valueOf(i)
            if (n.mod(divisor) == BigInteger.ZERO) {
                factors.add(divisor)
            }
        }

        return factors
    }

    fun factorRandom(): String {
        val randomNum = Random.nextInt(1, 26)
        val factors = factor(BigInteger.valueOf(randomNum.toLong()))
        return "Factor[$randomNum] = $factors"
    }

    fun primeFactor(n: BigInteger): List<BigInteger> {
        if (n == BigInteger.ZERO) {
            return emptyList()
        }

        val primeFactors = mutableListOf<BigInteger>()
        var num = n
        val two = BigInteger.valueOf(2)

        while (num.mod(two) == BigInteger.ZERO) {
            primeFactors.add(two)
            num = num.divide(two)
        }

        var i = 3L
        while (BigInteger.valueOf(i * i) <= num) {
            val divisor = BigInteger.valueOf(i)
            while (num.mod(divisor) == BigInteger.ZERO) {
                primeFactors.add(divisor)
                num = num.divide(divisor)
            }
            i += 2
        }

        if (num > BigInteger.ONE) {
            primeFactors.add(num)
        }

        return primeFactors
    }

    fun primeFactorRandom(): String {
        val randomNum = Random.nextInt(1, 26)
        val primeFactors = primeFactor(BigInteger.valueOf(randomNum.toLong()))
        return "PFactor[$randomNum] = $primeFactors"
    }
}

File: cabal/packages/numbers-utils/src/Numbers/Utils/Operations/FactorUtils.hs
module Numbers.Utils.Operations.FactorUtils (factor, factorRandom, primeFactor, primeFactorRandom) where

import Numeric.Natural (Natural)
import System.Random (randomRIO)

factor :: Natural -> [Natural]
factor n = [x | x <- [1 .. n], n `mod` x == 0]

factorRandom :: IO String
factorRandom = do
    randomNum <- randomRIO (1, 500000 :: Integer)
    let factors = factor (fromIntegral randomNum)
    return $ "Factor[" ++ show randomNum ++ "] = " ++ show factors

primeFactor :: Natural -> [Natural]
primeFactor 0 = []
primeFactor 1 = []
primeFactor num = go num 2
    where
        go n divisor
            | divisor * divisor > n = [n]
            | n `mod` divisor == 0 = divisor : go (n `div` divisor) divisor
            | divisor == 2 = go n (divisor + 1)
            | otherwise = go n (divisor + 2)

primeFactorRandom :: IO String
primeFactorRandom = do
    randomNum <- randomRIO (1, 500000 :: Integer)
    let primeFactors = primeFactor (fromIntegral randomNum)
    return $ "PFactor[" ++ show randomNum ++ "] = " ++ show primeFactors
